pkg/optimization/device_perf_linux.go

```diff
@@ -43,79 +45,105 @@ func (deviceHelper *DeviceHelper) DiskSupportsPerfOptimization(diskPerfProfile,
+}
 
-	err = echoToFile("1", filepath.Join(blockDeviceRootPath, deviceName, "queue/iosched/writes_starved"))
+func getDeviceSettingsForBasicProfile(nodeInfo *NodeInfo,
+	deviceRoot, perfProfile, accountType, diskSizeGibStr, diskIopsStr, diskBwMbpsStr string) (deviceSettings map[string]string, err error) {
+	klog.V(2).Infof("getDeviceSettingsForBasicProfile: Getting settings for deviceRoot %s",
+		deviceRoot)
+	queueDepth, nrRequests, scheduler, maxSectorsKb, _, err := getOptimalDeviceSettings(nodeInfo, DiskSkuMap, perfProfile, accountType, diskSizeGibStr, diskIopsStr, diskBwMbpsStr)
 	if err != nil {
-		return fmt.Errorf("OptimizeDiskPerformance: Could not set queue/iosched/writes_starved for device %s. Error: %v", deviceName, err)
+		return nil, fmt.Errorf("getDeviceSettingsForBasicProfile: Failed to get optimal settings for profile %s accountType %s. Error: %v", perfProfile, accountType, err)
 	}
 
-	err = echoToFile(queueDepth, filepath.Join(blockDeviceRootPath, deviceName, "device/queue_depth"))
-	if err != nil {
-		return fmt.Errorf("OptimizeDiskPerformance: Could not set queue/queue_depth for device %s. Error: %v", deviceName, err)
-	}
+	deviceSettings = make(map[string]string)
+	deviceSettings[filepath.Join(deviceRoot, "queue/max_sectors_kb")] = maxSectorsKb
+	deviceSettings[filepath.Join(deviceRoot, "queue/scheduler")] = scheduler
+	deviceSettings[filepath.Join(deviceRoot, "device/queue_depth")] = queueDepth
+	deviceSettings[filepath.Join(deviceRoot, "queue/nr_requests")] = nrRequests
+	deviceSettings[filepath.Join(deviceRoot, "queue/read_ahead_kb")] = "8"
 
-	err = echoToFile(nrRequests, filepath.Join(blockDeviceRootPath, deviceName, "queue/nr_requests"))
-	if err != nil {
-		return fmt.Errorf("OptimizeDiskPerformance: Could not set queue/nr_requests for device %s. Error: %v", deviceName, err)
-	}
+	return deviceSettings, nil
+}
 
-	err = echoToFile(readAheadKb, filepath.Join(blockDeviceRootPath, deviceName, "queue/read_ahead_kb"))
-	if err != nil {
-		return fmt.Errorf("OptimizeDiskPerformance: Could not set queue/read_ahead_kb for device %s. Error: %v", deviceName, err)
+func getDeviceSettingsForAdvancedProfile(deviceRoot string, deviceSettingsFromCtx map[string]string) (deviceSettings map[string]string, err error) {
+	klog.V(2).Infof("getDeviceSettingsForAdvancedProfile: Getting settings for deviceRoot %s deviceSettingsFromCtx %v",
+		deviceRoot,
+		deviceSettingsFromCtx)
+	deviceSettings = make(map[string]string)
+	for setting, value := range deviceSettingsFromCtx {
+		deviceSettings[filepath.Join(deviceRoot, setting)] = value
 	}
 
-	err = echoToFile("0", filepath.Join(blockDeviceRootPath, deviceName, "queue/wbt_lat_usec"))
-	if err != nil {
-		return fmt.Errorf("OptimizeDiskPerformance: Could not set queue/wbt_lat_usec for device %s. Error: %v", deviceName, err)
+	return deviceSettings, nil
+}
+
+func updateDeviceSettings(deviceRoot string, deviceSettings map[string]string) (err error) {
```

Comment:
Consider `applyDeviceSettings`. To me `updateDeviceSettings` seems like we're updating the PV parameters.